### PR TITLE
Filters

### PR DESCRIPTION
**Pull request by fdansv**

### CODE REVIEW

File: src/renderer.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ var cartodb = global.cartodb || {};
 var carto = global.carto || require('carto');
 var _ = global._ || require('underscore');
 var geo = require("./geo");
+var Filter = require("./filter");
```
Comment (pabletecodes) [inline src/renderer.js:6]: I don't understand why the Filter is part of the renderer, and why is renderer responsible of adding tiles to the Filter. I'd expect someone listening to the "tileAdded" event to load the data into an object where you can add filters.

**Comment by pabletecodes** [comment]
I've been looking into this a bit and I still don't quite get the concept of Filter and Expressions. Perhaps it's because of the naming... 

I would expect to have an object where you could add filters and query for the stuff we need for widgets. Here's some pseudocode (I called this object datasource but it could be whatever makes sense)

```

// { city: 'Madrid', population: '10' }
// { city: 'Barcelona', population: '20' }
// { city: 'A Coruña', population: '30' }

var categoryFilter = new CategoryFilter({
  widgetId: 'myWidget1', // not sure if this is needed
  column: 'city'
});
var rangeFilter = new RangeFilter({
  widgetId: 'myWidget2', // not sure if this is needed
  column: 'population',
  bins: 10
});

datasource.addFilter(rangeFilter);
datasource.addFilter(categoryFilter);

datasource.getCategories('myWidget1');
// It would return all cities

// Change the filter
rangeFilter.setRange(20, 30);

datasource.getCategories(); // not sure if it would need more paremeters like the column or widgetId
// It would return "Barcelona" and "A Coruña"

categoryFilter.reject("Barcelona")

datasource.getCategories(); // not sure if it would need more paremeters like the column or widgetId
// It would return "A Coruña"

datasource.getHistogram();
```

**Comment by fdansv** [comment]
@alonsogarciapablo right now I'm focusing more on the logic of the widgets themselves, but here's the way I'm seeing it:
- First of all, I called expressions to the declarations of widgets within the filtering system. It's probably a poor choice of naming and I'm open to changing it, or just call them widgets as well. For the sake of clarity I'll refer to them here as expressions.
- When we add a new expression, we're asking the Filter to "keep track" of the column, with the parameters indicated. The main reason why we do this is that we have to instantiate a [`dimension`](https://github.com/square/crossfilter/wiki/API-Reference#dimension) with it. Dimensions are super expensive to create and store, so we can't just create new ones as the widgets ask for data. 

``` javascript
filter.addExpression("category_de_pepito", {
  type: "aggregation",
  column: "sov_a3",
  aggregation: "count",
  sync: true
});
```
- When we keep track of that column with those parameters, this means that their returning data will be recalculated each time `filter.update()` is run. What that method does is iterating over the list of expressions that have been declared and run the function that was generated for each one at the moment of the declaration (see [this](https://github.com/CartoDB/d3.cartodb/pull/42/files#diff-d2ddee121ba21440f473010082e89e72R80), for example).

``` javascript
filter.update();
// will update the this.report object, and return:

{
  "category_de_pepito": {
    "count": 2722,
    "nulls": 0,
    "min": 1,
    "max": 145,
    "categoriesCount": 136,
    "categories": [
      {
        "category": "RUS",
        "value": 145,
        "agg": false
      }, // [...]
    ],
    "type": "aggregation"
  }
}

```
- As soon as I'm done coding histograms I'll add both a method to _filter according_ the widget id and a getter function for the widgets to use.

**Comment by javisantana** [comment]
I think the abstraction level for that api is higher than expected. I'd expect something like:

for filtering:

```
layer.filterRange('attr', [0, 1000]);
layer.filterExact('attr2', 'pepe');
layer.clearFilters();
```

and to fetch data:

```
var values = layer.getValues('attr') // values will contain the attr values with the current filters applied
// histogram other stuff is calculated outside the layer
```

**Comment by pabletecodes** [comment]
Thanks for your explanation @fdansv! :+1: Here're some comments/questions:

> First of all, I called expressions to the declarations of widgets within the filtering system. It's probably a poor choice of naming and I'm open to changing it, or just call them widgets as well. For the sake of clarity I'll refer to them here as expressions.

IMO what we are doing here is adding filters to an object that contains data. Each (not all) widget have a filter. I propose renaming "expressions" to "filters".

> When we add a new expression, we're asking the Filter to "keep track" of the column, with the parameters indicated. The main reason why we do this is that we have to instantiate a dimension with it. Dimensions are super expensive to create and store, so we can't just create new ones as the widgets ask for data.

This makes sense to me :+1: Although I still think that it's probably the information about the filter what we need to keep track of (not the info about the widget). For example, for an histogram widget, we'd basically need to keep track of a filter of type "range" that is acting on an "X" column and has the "min" and "max" options. Does this make sense?

> When we keep track of that column with those parameters, this means that their returning data will be recalculated each time filter.update() is run. What that method does is iterating over the list of expressions that have been declared and run the function that was generated for each one at the moment of the declaration (see this, for example).

Instead of having to call `.update` explicitly, would it make sense to "automagically" do this internally when, for example, a new "expression" is added? Or every time we need new data?

> As soon as I'm done coding histograms I'll add both a method to filter according the widget id and a getter function for the widgets to use.

Great. That's the part I'm missing... I still don't know how you would reject one category or all categories, or change the set the max and min of a range filter (histogram). Perhaps you can elaborate a bit more on this.

Thanks!

**Comment by pabletecodes** [comment]
Had a quick chat with @javisantana, and we came up with a bunch of more clearer requirements for L.CartoDB3Layer:
- Since this layer could be used in a standalone way, it shouldn't know anything about widgets.
- (related to :point_up: ) this type of layer should not know how to generate the data for the widgets.
- This layer is mainly responsible of:
  - Rendering a bunch of features from the specified provider.
  - Provide a mechanism to filter the features that are rendered (and the data :point_down: ).
  - Provide a mechanism to query the data that has been loaded by the provider (eg: the data contained in the GeoJSON tiles).

cc: @fdansv

**Comment by fdansv** [comment]
OK, so as agreed I'm focusing on the filtering part of the stack, leaving the analysis to cartodb.js. I've added the following API elements:
- `filterRange`
- `filterAccept`
- `filterReject`
- `clearFilters`
- `clearFilter`
- `getValues`

These are all functional so feel free to test. Any comments?

**Comment by fdansv** [comment]
OK so now the `Filter` is connected to the rendering part so setting a filter with the API above will reflect those changes on the map, which will update when `renderer.redraw()` is called. Also I added a few tests to provide good coverage for filters.

@rochoa @alonsogarciapablo this is ready to merge my side.

File: src/filter.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
+var Crossfilter = require("crossfilter");
+
+function Filter(){
+  this.crossfilter = new Crossfilter();
+  this.dimensions = {};
+  this.tiles = new Set();
```
Comment (rochoa) [inline src/filter.js]: Set is a ES6 feature. We need another approach or a polyfill.

File: src/tileloader.js [hunk 3]
```diff
@@ -55,9 +55,9 @@ module.exports = L.Class.extend({
       for (i = bounds.min.x; i <= bounds.max.x; i++) {
         point = new L.Point(i, j);
         point.zoom =  zoom;
-        if (this._tileShouldBeLoaded(point)) {
+        // if (this._tileShouldBeLoaded(point)) {
```
Comment (rochoa) [inline src/tileloader.js]: Shouldn't we remove this?

File: src/renderer.js (hunk 1 above)
Comment (rochoa) [inline src/renderer.js:6]: For me it's a bit strange as well.

File: src/providers/windshaft.js [hunk 4]
```diff
@@ -1,13 +1,76 @@
...
+	      }.bind(this));
+	    }
+	  }
+  },
+
+  getGeometry: function(url, callback){
+      d3.json(url, callback);
+  },
+
+	_generateMapconfig: function(table) {
```
Comment (rochoa) [inline src/providers/windshaft.js]: When do we need this?

File: spec/filter.spec.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
+describe("The filter", function(){
+	beforeAll(function(){
+		this.filter = new cartodb.d3.filter();
+		this.tile = MOCK_TILE;
```
Comment (rochoa) [inline spec/filter.spec.js:4]: Where is MOCK_TILE defined?

**Comment by rochoa** [comment]
As per previous comments I was expecting to have the filtering at layer level, but probably I'm missing something here.

Having some kind of integration test would help a lot to understand the layer high level API.

File: src/renderer.js (hunk 1 above)
Comment (fdansv) [inline src/renderer.js:6]: The layer (`CartoDBd3Layer`) receives a geojson tile that contains/can contain several layers, each one of which is represented by a renderer. Keeping that in mind, would it be better to create a new class, called something along the lines of `sublayer`, and restrict the functions of the renderer to generating the svgs?

File: spec/filter.spec.js (hunk 5 above)
Comment (fdansv) [inline spec/filter.spec.js:4]: `spec/mocks/tile.js`

File: src/providers/windshaft.js (hunk 4 above)
Comment (fdansv) [inline src/providers/windshaft.js]: We need it to use the library as a standalone library outside cartodb.js, whilst also avoiding the use of the SQL API. Similar thing to what we do in Torque.

File: src/filter.js (hunk 2 above)
Comment (fdansv) [inline src/filter.js]: :ok_hand: just changed this, it was supposed to be a temporary use. Replaced it with a normal JS object.

File: src/providers/windshaft.js [hunk 6]
```diff
@@ -1,13 +1,76 @@
...
 	this.user = options.user;
 	this.table = options.table;
 	this.format = options.format;
+	this.options = options;
 	this.tileCache = {};
+	this.initialize();
 }
 
 WindshaftProvider.prototype = {
+	initialize: function(){
```
Comment (rochoa) [inline src/providers/windshaft.js]: Why don't we have this in the constructor? Is there any other way to use initialize apart from being called from the constructor?

File: src/providers/windshaft.js (hunk 4 above)
Comment (rochoa) [inline src/providers/windshaft.js]: Oh, I see.

File: src/providers/windshaft.js [hunk 7]
```diff
@@ -1,13 +1,76 @@
...
+		var self = this;
+		this.tiler_template = this.tiler_template.replace("{user}", this.user);
+		var mapconfig = this._generateMapconfig(this.table);
+		var url = this.tiler_template + "/api/v1/map?config=" + encodeURIComponent(JSON.stringify(mapconfig));
+		cartodb.d3.net.jsonp(url + "&callback=mapconfig", function(data){
+			self.layergroup = data;
```
Comment (rochoa) [inline src/providers/windshaft.js]: I see a mix of tabs/spaces for indenting. Can we be a bit more consistent? :smiling_imp:

File: src/filter.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  clearFilter: function(column) {
+    this.dimensions[column].filterAll();
+  },
+
+  getValues: function(column) {
+    return this.dimensions[column].top(Infinity);
+  },
+
+  // setBoundingBox: function(north, east, south, west) {
```
Comment (rochoa) [inline src/filter.js]: Remove it.

File: src/filter.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    this.dimensions[column].filter(function(f){
+      if (terms.indexOf(f) > -1){
+        return true;
+      }
+    });
+
+  },
+
+  filterReject: function(column, terms){
```
Comment (rochoa) [inline src/filter.js:61]: All filter\* functions are quite similar. Wouldn't be better to have a generic filter function that receives the actual filter function and applies it by column?

Or even better, having different filter entities with a common interface that can be applied. Having different functions for different types of filters means that we need as many public functions as filters we want to have.

File: src/providers/windshaft.js (hunk 7 above)
Comment (fdansv) [inline src/providers/windshaft.js]: Yes sorry, sublime does this for some reason. It drives me crazy too. Correcting...

**Comment by rochoa** [comment]
I'm still missing some tests or examples displaying how the layer/high level API is.

**Pull request merged**